### PR TITLE
8234773: Fix ThreadsSMRSupport::_bootstrap_list

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to ThreadsList and the singleton _bootstrap_list.
ThreadsList is made noncopyable, with _bootstrap_list changed to be direct
initialized to avoid referencing the copy constructor.  The ThreadsList
constructor now uses a static array for the 0-entry case, so that static
ctor/dtor of _bootstrap_list doesn't involve C-heap allocation.

Testing:
mach5 tier1

Local (linux-x64) build with -fno-elide-constructors and ran hotspot:tier1.
This has a failure unrelated to this change:
compiler/intrinsics/klass/CastNullCheckDroppingsTest.java fails with
#  Internal Error (../../src/hotspot/share/jfr/utilities/jfrVersionSystem.inline.hpp:98), pid=31086, tid=31713
#  assert(node->_live) failed: invariant
Filed new bug: https://bugs.openjdk.java.net/browse/JDK-8259036

/label hotspot-runtime
/summary Make ThreadsList noncopyable, direct initializing _bootstrap_list. Avoid C-heap allocation for _bootstrap_list.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8234773](https://bugs.openjdk.java.net/browse/JDK-8234773): Fix ThreadsSMRSupport::_bootstrap_list

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to d3240a187a05d183609be6df451cc2c5da14d5b5
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to 1a6cbcb93dfbc7dc5121fa072d420522ba790b1b

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1921/head:pull/1921`
`$ git checkout pull/1921`

### CODE REVIEW

File: src/hotspot/share/runtime/threadSMR.cpp [hunk 1]
```diff
@@ -76,7 +76,7 @@ volatile uint         ThreadsSMRSupport::_deleted_thread_time_max = 0;
 volatile uint         ThreadsSMRSupport::_deleted_thread_times = 0;
 
 // The bootstrap list is empty and cannot be freed.
-ThreadsList ThreadsSMRSupport::_bootstrap_list = ThreadsList(0);
+ThreadsList ThreadsSMRSupport::_bootstrap_list{0};
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/threadSMR.cpp:79]: To anyone unfamiliar with the details of direct initialization this is very obscure. I don't know how to interpret this given a ThreadsList has four member fields. Wouldn't `{}` have the same effect of implicitly zero-initializing all the fields?
Stylistically I find this syntax jarring, an equals sign in there would look far more natural to me. :(

**Review by dholmes-ora — APPROVED** [review]
Hi Kim,
Overall the fix looks good. I have one comment on the direct initialization part (after reading up on direct initialization).

Thanks,
David

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.
Like @dholmes-ora, I'm not familiar with this "explicit" stuff.

**Comment by dholmes-ora** [comment]
I'm now more familiar with the use of "explicit" so that part is fine too.

I'd prefer to use this syntax if it is applicable to this case:

ThreadsList ThreadsSMRSupport::_bootstrap_list(0);

as the `{ 0 }` syntax just looks weird to me as an implicit constructor invocation. Your call.

Thanks,
David

**Pull request closed without merging**